cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -0,0 +1,145 @@
+  Unknown,
+  // Target is a regular file
+  File,
+  // Target is a directory
+  Directory
+};
+
+std::string ToString(FileType);
+
+struct FileStats {
+  FileType type = FileType::Unknown;
+  // The file path in the filesystem.
+  std::string path;
+  // The size in bytes, if available.  Only regular files are guaranteed
+  // to have a size.
+  int64_t size = -1;
+  // The time of last modification, if available.
+  TimePoint mtime = TimePoint(-1);
+
+  FileStats() = default;
+
+  // The file base name (component after the last '/').
+  std::string base_name() const;
+};
+
+struct OpenFlags {
+  // Whether to transparently compress or decompress, based on file extension
+  // or other heuristics.
+  bool transparent_compression = true;
+
+  OpenFlags() = default;
+};
+
+struct Selector {
+  // The folder in which to select files.
+  std::string base_folder;
+  // Whether to recurse into subfolders.
+  bool recursive = false;
+
+  MatchOptions() = default;
+};
+
+
+class FileSystem {
+ public:
+  virtual ~FileSystem();
+
+  /// Get statistics for the given target.
+  ///
+  /// Any symlink is automatically dereferenced, recursively
+  virtual Status GetTargetStats(const std::string& path, FileStats* out);
+  /// Same, for many targets at once.
+  virtual Status GetTargetStats(const std::vector<std::string>& paths,
+                                std::vector<FileStats>* out) = 0;
+  virtual Status GetTargetStats(const Selector& select,
+                                std::vector<FileStats>* out) = 0;
+
+  /// Create a folder and subfolders.
+  virtual Status CreateFolder(const std::string& path, bool recursive = true) = 0;
+
+  /// Delete a folder and contents.
+  virtual Status DeleteFolder(const std::string& path) = 0;
+
+  /// Delete a file.
+  virtual Status DeleteFile(const std::string& path) = 0;
+  /// Delete many files.
+  virtual Status DeleteFiles(const std::vector<string>& path) = 0;
+
+  /// Move / rename a file or directory.
+  ///
+  /// The destination will be replaced if it exists.
+  virtual Status Move(const std::string& src, const std::string& dest) = 0;
+
+  /// Copy a file.
+  ///
+  /// The destination will be replaced if it exists.
+  virtual Status Copy(const std::string& src, const std::string& dest) = 0;
+
+  /// Open an input stream.
+  virtual Status OpenInputStream(const std::string& path,
+                                 std::shared_ptr<InputStream>* out);
+
+  virtual Status OpenInputStream(const std::string& path, OpenFlags flags,
+                                 std::shared_ptr<InputStream>* out) = 0;
+
+  /// Open an output stream.
+  ///
+  /// If the target already exists, it is replaced with a new file.
+  virtual Status OpenOutputStream(const std::string& path,
+                                  std::shared_ptr<OutputStream>* out);
+
+  virtual Status OpenOutputStream(const std::string& path, OpenFlags flags,
+                                  std::shared_ptr<OutputStream>* out) = 0;
+};
```

Comment:
May want to add some APIs about opening memory mapped files (similar to https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/file_system.h#L92), as well as probably an API indicating whether this feature is supported or not. 

What about `RandomAccessFile`?

Reply:
I'm not sure what the point of a specific memory-mapping API would be. Why would you fail if memory mapping is not possible?
(of course, it's always possible to copy the file locally and then memory map it... but it may not be a terrific idea if the file is huge)

Same question for `RandomAccessFile`: what do we need it for, in addition to `InputStream`?

Reply:
For Dask, in particular, it is essential to be able to open a file and read only some portion of it without scanning through from the start. When there is also compression, this may of may not be possible, depending on the compression. Did I misunderstand?

Reply:
Can you explain in which situations it is essential?

Reply:
CSV file on S3 is 90GB large, and the current task on the current worker wishes to interpret the final 100MB into a dataframe. It had better not have to read through the first 89.9GB of the data to be able to do that

Reply:
I see, so we have to support that.

Reply:
Right, so this amounts to `Bytes` range requests for S3/gcsfs/azure, but might be more complex for other file-systems (e.g., I also implemented SSH, FTP, HTTP); and as I alluded to above, may also be possible in the presence of compression, if the format is block-wise (like xz).

Reply:
@pitrou file formats like Parquet rely on random access

Reply:
I see :-/

Reply:
memory mapped is also mandatory for zero-copy (or minimizing the copies by avoiding to copy from kernel buffer cache to user space pages). memory mapping huge files shouldn't be an issue on Linux.

Reply:
The question is whether there should be an explicit API for memory-mapped files (which will fail on most filesystems and require a fallback in user code) or it's purely an internal optimization.

Reply:
We could possibly adopt something similar to TensorFlow -- "use memory map if available, otherwise read file into memory and return BufferReader"